File: typescript/tests/payload_schema_crosscheck.test.ts
/**
 * Dev-only third-party cross-check for the payload-schema validator
 * (effects contract §19.5).
 *
 * `@hyperjump/json-schema` is a devDependency and never a runtime one: it
 * exists to assert that the in-house validator's verdicts agree with an
 * independent implementation on every shared vector. A disagreement is a test
 * failure to investigate, never something the code resolves for itself.
 *
 * Two families are excluded by construction, because they are ours and not
 * JSON Schema's: decision 16's magnitude guard, and JSON representability
 * (which a JSON vector file cannot express in the first place).
 *
 * The declared literal carries no `$schema` -- the closed subset rejects that
 * keyword -- so the dialect is spliced onto a COPY here, purely to tell
 * hyperjump which dialect to compile against.
 */

import { strict as assert } from "node:assert";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { test } from "node:test";
import { fileURLToPath } from "node:url";
import { registerSchema, validate } from "@hyperjump/json-schema/draft-2020-12";
import { validatePayloadValue } from "../src/payload_schema.js";

const vectorsPath = join(
	dirname(fileURLToPath(import.meta.url)),
	"..",
	"..",
	"..",
	"conformance",
	"payload_schema_vectors.json",
);

interface Vector {
	readonly name: string;
	readonly schema: Record<string, unknown>;
	readonly value?: unknown;
	readonly valid: boolean;
	readonly detail?: string;
	readonly unrepresentable_in?: readonly string[];
}

interface VectorDoc {
	readonly instance_vectors: readonly Vector[];
}

const doc = JSON.parse(readFileSync(vectorsPath, "utf8")) as VectorDoc;

/** The two details JSON Schema itself has no opinion about. */
const OURS_ONLY = new Set([
	"the number's magnitude exceeds 2^53 (declare a big identifier as a string)",
	"the value is not representable in JSON",
]);

function crossCheckable(v: Vector): boolean {
	if ((v.unrepresentable_in ?? []).includes("typescript")) {
		return false;
	}
	return v.valid || !OURS_ONLY.has(v.detail ?? "");
}

const checkable = doc.instance_vectors.filter(crossCheckable);

test("the cross-check covers most of the instance matrix", () => {
	assert.ok(checkable.length >= 130, `only ${checkable.length} vectors`);
});

let uriSeq = 0;

for (const v of checkable) {
	test(`cross-check: ${v.name}`, async () => {
		uriSeq += 1;
		const uri = `https://strictcli.test/payload-schema/${uriSeq}`;
		registerSchema(
			{
				$schema: "https://json-schema.org/draft/2020-12/schema",
				...v.schema,
			},
			uri,
		);
		const theirValidator = await validate(uri);
		// The instance type is spelled through the validator's own signature
		// rather than imported from a transitive package.
		type Instance = Parameters<typeof theirValidator>[0];
		const theirs = theirValidator(v.value as Instance).valid;
		const ours = validatePayloadValue(v.value, v.schema) === null;
		assert.equal(
			ours,
			theirs,
			`in-house says ${ours}, hyperjump says ${theirs}`,
		);
	});
}

File: go/strictcli/payload_schema_crosscheck_test.go
package strictcli

// Dev-only third-party cross-check for the payload-schema validator
// (effects contract §19.5).
//
// santhosh-tekuri/jsonschema v6 is a TEST dependency and never a runtime one:
// it exists to assert that the in-house validator's verdicts agree with an
// independent implementation on every shared vector. A disagreement is a test
// failure to investigate, never something the code resolves for itself.
//
// Two families are excluded by construction, because they are ours and not
// JSON Schema's: decision 16's magnitude guard, and JSON representability
// (which a JSON vector file cannot express in the first place).

import (
	"testing"

	"github.com/santhosh-tekuri/jsonschema/v6"
)

// crossCheckable reports whether a vector's verdict is one JSON Schema itself
// decides.
func crossCheckable(v payloadVector) bool {
	if v.skipHere() {
		return false
	}
	if v.Valid {
		return true
	}
	return v.Detail != pdetailMagnitude && v.Detail != pdetailNotJSON
}

func compileVectorSchema(t *testing.T, schema interface{}) *jsonschema.Schema {
	t.Helper()
	c := jsonschema.NewCompiler()
	if err := c.AddResource("vector.json", schema); err != nil {
		t.Fatalf("add resource: %v", err)
	}
	sch, err := c.Compile("vector.json")
	if err != nil {
		t.Fatalf("compile: %v", err)
	}
	return sch
}

func TestPayloadValidatorAgreesWithThirdParty(t *testing.T) {
	doc := loadPayloadVectors(t)
	checked := 0
	for _, v := range doc.InstanceVectors {
		if !crossCheckable(v) {
			continue
		}
		checked++
		t.Run(v.Name, func(t *testing.T) {
			schemaTree := decodeVectorJSON(t, v.Schema)
			schema := decodeVectorSchema(t, v.Schema)
			value := decodeVectorJSON(t, v.Value)
			theirs := compileVectorSchema(t, schemaTree).Validate(value) == nil
			ours := validatePayloadValue(value, schema) == nil
			if ours != theirs {
				t.Fatalf("in-house says %v, santhosh-tekuri v6 says %v", ours, theirs)
			}
		})
	}
	if checked < 130 {
		t.Fatalf("cross-check covered only %d vectors", checked)
	}
}

func TestAcceptedSchemasCompileAsJSONSchema(t *testing.T) {
	// Anything the subset admits must be a legal JSON Schema document: the
	// subset is a restriction of JSON Schema, never a dialect of its own.
	doc := loadPayloadVectors(t)
	for _, v := range doc.SchemaVectors {
		if !v.Valid || v.skipHere() {
			continue
		}
		t.Run(v.Name, func(t *testing.T) {
			compileVectorSchema(t, decodeVectorJSON(t, v.Schema))
		})
	}
}
